File: Extras/pytest/tests/test_04_Functional.py
import pytest
import websockets
import unium

@pytest.mark.asyncio
async def test_webSocket( unium_socket ):

  """
  a functional test
  run around the tutorial level and checks we can collect all the pickups
  """

  # connect to the game
  async with websockets.connect( unium_socket ) as ws:

    u = unium.WebsocketHelper( ws )

    # check we have the unium project 
    about = await u.get( "/about" )
    assert about[ "Product" ] == "unium"

    # (re)load the tutorial scene (ensures it's in the start state)
    res = await u.get( "/utils/scene/Tutorial" )
    assert res[ "scene" ] == "Tutorial"

    # get positions of all pickups
    pickups = await u.get( "/q/scene/Game/Pickup.transform.position" )

    # we should have 4 to start with
    assert len( pickups ) == 4

    # register for OnPickupCollected events
    await u.bind( "/scene/Game.Tutorial.OnPickupCollected" )

    # move to each pickup
    for pos in pickups:
      await u.get( f"/q/scene/Game/Player.Player.MoveTo({pos})" )
      await u.wait_for( "OnPickupCollected", timeout=10.0 )

    # check there are no more pickups left in the level
    pickups = await u.get( "/q/scene/Game/Pickup.name" )
    assert len( pickups ) == 0

File: Extras/nunit/Test04_Functional.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using NUnit.Framework;
using System.Threading.Tasks;
using Unium.Helpers;

namespace Unium.Test
{
    [TestFixture]
    public class Test04_Functional
    {
        [Test]
        public async Task CollectPickups()
        {
            using( var u = new WebsocketHelper() )
            {
                // connect to game

                await u.Connect( TestConfig.WS );


                // check this is the unium project running

                dynamic about = await u.Get( "/about" );

                Assert.AreEqual( "gwaredd", (string) about.Company );
                Assert.AreEqual( "unium", (string) about.Product );
                Assert.AreEqual( "Tutorial", (string) about.Scene );


                // reload the tutorial scene (ensures we are in the start state)

                dynamic scene = await u.Get( "/utils/scene/Tutorial" );

                Assert.IsNotNull( scene.scene );
                Assert.AreEqual( "Tutorial", (string) scene.scene );


                // get positions of all pickups

                dynamic pickups = await u.Get( "/q/scene/Game/Pickup.transform.position" );

                Assert.IsNotNull( pickups.Count == 4 );

                // collect pickups

                await u.Bind( "/scene/Game.Tutorial.OnPickupCollected" );

                foreach( JToken pickup in pickups )
                {
                    var pos = pickup.ToString( Formatting.None );

                    await u.Get( $"/q/scene/Game/Player.Player.MoveTo({pos})" );
                    await u.WaitFor( "OnPickupCollected", 10.0f );
                }


                // check there are no more pickups left in the level

                dynamic pickupsLeft = await u.Get( "/q/scene/Game/Pickup.name" );
                Assert.IsNotNull( pickupsLeft.Count == 0 );


                // diconnect

                await u.Disconnect();
            }
        }
    }
}
